File: day15.rb
# STARTING_VALUES = [65, 8921].freeze
STARTING_VALUES = [591, 393].freeze
FACTORS = [16_807, 48_271].freeze
DIVISOR = 2_147_483_647
N = 4 * 10**7

def part1(starting_values, factors)
  starting_values.zip(factors)
                 .map { |a, b| generate_all(a, b) }
                 .transpose
                 .count { |a, b| match?(a, b) }
end

def generate_all(starting_value, factor)
  N.times
   .reduce([starting_value]) { |a, _e| a.push(generate(a.last, factor)) }
   .drop(1)
end

def generate(prev, factor)
  prev * factor % DIVISOR
end

def match?(a, b)
  a % 2**16 == b % 2 **16
end

if __FILE__ == $PROGRAM_NAME
  p part1(STARTING_VALUES, FACTORS)
end

File: haskell/day15.hs
--startingValues = [65, 8921]
startingValues = [591, 393]
factors = [16807, 48271]
d = [4, 8]
divisor = 2147483647
round2to16 = round $ 2**16
numberOfPairs = round $ 4 * 10 ** 7
numberOfPairs2 = round $ 5 * 10 ** 6

part1 n = length [p | p <- take n pairs, match p]

part2 n = length [p | p <- take n pairs2, match p]

pairs = zip (xs !! 0) (xs !! 1)
    where xs = map gen ys
          ys = zip factors startingValues

pairs2 = zip (xs !! 0) (xs !! 1)
  where xs = map gen2 ys
        ys = zip3 factors startingValues d

gen (factor, startingValue) = iterate (next factor) startingValue

gen2 (factor, startingValue, d) = filter (\x -> rem x d == 0) xs
    where xs = gen (factor, startingValue)

next factor x = rem (x * factor) divisor

match (a, b) = rem a round2to16 == rem b round2to16

main = do
   print (part1 numberOfPairs)
   print (part2 numberOfPairs2)
